FreeRTOS_Sockets.c

```diff
@@ -1878,371 +1980,539 @@ void * vSocketClose( FreeRTOS_Socket_t * pxSocket )
-                                   break;
-                           }
-                       }
+/**
+ * @brief Handle the socket option FREERTOS_SO_WIN_PROPERTIES, which sets
+ *        the sizes of the TCP windows and the sizes of the stream buffers.
+ *
+ * @param[in] pxSocket: The socket whose options are being set.
+ * @param[in] pvOptionValue: The pointer that is passed by the application.
+ */
+    static BaseType_t prvSetOptionTCPWindows( FreeRTOS_Socket_t * pxSocket,
+                                              const void * pvOptionValue )
+    {
+        BaseType_t xReturn = -pdFREERTOS_ERRNO_EINVAL;
+        const WinProperties_t * pxProps;
 
-                        xReturn = 0;
-                        break;
-                #endif /* ipconfigUSE_CALLBACKS */
+        do
+        {
+            if( pxSocket->ucProtocol != ( uint8_t ) FREERTOS_IPPROTO_TCP )
+            {
+                FreeRTOS_debug_printf( ( "Set SO_WIN_PROP: wrong socket type\n" ) );
+                break; /* will return -pdFREERTOS_ERRNO_EINVAL */
+            }
 
-                #if ( ipconfigUSE_TCP != 0 )
-                    #if ( ipconfigSOCKET_HAS_USER_SEMAPHORE != 0 )
+            if( ( pxSocket->u.xTCP.txStream != NULL ) || ( pxSocket->u.xTCP.rxStream != NULL ) )
+            {
+                FreeRTOS_debug_printf( ( "Set SO_WIN_PROP: buffer already created\n" ) );
+                break; /* will return -pdFREERTOS_ERRNO_EINVAL */
+            }
 
-                        /* Each socket has a semaphore on which the using task normally
-                         * sleeps. */
-                        case FREERTOS_SO_SET_SEMAPHORE:
-                           {
-                               pxSocket->pxUserSemaphore = *( ipPOINTER_CAST( SemaphoreHandle_t *, pvOptionValue ) );
-                           }
-                            xReturn = 0;
-                            break;
-                    #endif /* ipconfigSOCKET_HAS_USER_SEMAPHORE */
+            pxProps = ipPOINTER_CAST( const WinProperties_t *, pvOptionValue );
 
-                    #if ( ipconfigSOCKET_HAS_USER_WAKE_CALLBACK != 0 )
-                        case FREERTOS_SO_WAKEUP_CALLBACK:
+            xReturn = prvSockopt_so_buffer( pxSocket, FREERTOS_SO_SNDBUF, &( pxProps->lTxBufSize ) );
 
-                            /* Each socket can have a callback function that is executed
-                             * when there is an event the socket's owner might want to
-                             * process. */
-                            /* The type cast of the pointer expression "A" to type "B" removes const qualifier from the pointed to type. */
-                            pxSocket->pxUserWakeCallback = ipCAST_CONST_PTR_TO_CONST_TYPE_PTR( SocketWakeupCallback_t, pvOptionValue );
-                            xReturn = 0;
-                            break;
-                    #endif /* ipconfigSOCKET_HAS_USER_WAKE_CALLBACK */
+            if( xReturn != 0 )
```

Comment:
This violates MISRA Rule 15.4 which advises against multiple break statements inside a loop.

Reply:
This falls under an exception documented in MISRA.md.